---
title: "Process CCLE RNASeq dataset"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary
This notebook maps different gene and protein identifiers to Entrez IDs. This is needed to get the maximum information out of my metabolic models.

# Load up libraries
First, let's load up some data science libraries to help manipulate some tables / dataframes.
```{r, warning=FALSE, message=FALSE}
#install.packages("tidyverse")
#install.packages("readxl")

library(readxl)
library(dplyr)
```
Let's also load up Bioconductor and the mapping database:
```{r, warning=FALSE, message=FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install(c("hgu133plus2.db", "org.Hs.eg.db"))
library(hgu133plus2.db)
library(org.Hs.eg.db)
```

# Load up the data that we'll be mapping
Next, we need to load up the data that we wish to map.

## Cancer Cell Line Encyclopedia
```{r, warning=FALSE, message=FALSE}
filename = "~/Data/RNASeq/CCLE/Processed_CCLE_Effect.csv"
effect = read.csv(filename)
filename = "~/Data/RNASeq/CCLE/Processed_CCLE_Pvalue.csv"
pvalue = read.csv(filename)
ensembl = sub("\\.\\d+$", "", effect[, 1])
```

## Khao et al., 2020
```{r, warning=FALSE, message=FALSE}
filename = "~/Data/RNASeq/Dou/mmc3.xlsx"
effect = read_excel(filename, sheet='Differential gene expression', skip=1)
print(effect)
```

## Lu et al., 2019
```{r, warning=FALSE, message=FALSE}
filename = "~/Data/Proteomics/Garcia/DEP_EMT.xlsx"
effect = read_excel(filename, sheet='Log2FC')
pvalue = read_excel(filename, sheet='PValue')

print(effect)
```

# Map the identifiers

## Get gene ids to map

### Khao et al., 2020
```{r, warning=FALSE, message=FALSE}
mouse_symbol = apply(effect[, 1], 2, toupper)
```

### Lu et al., 2019
```{r, warning=FALSE, message=FALSE}
mouse_symbol = apply(effect[, 3], 2, toupper)
```

Finally, I've created several functions that will allow us to map identifiers using any dataset. 
```{r, warning=FALSE, message=FALSE}
map_symbol_to_entrez = function(symbols, keyType){
  # Create a map of the probes to Entrez and gene symbols
  map = AnnotationDbi::select(hgu133plus2.db,
               keytype=keyType,
               keys=symbols, 
               use.names=FALSE, 
               columns = c('ENTREZID', 'SYMBOL')
               )
  return(map)
}
map = map_symbol_to_entrez(mouse_symbol, "SYMBOL")
print(map)
```

## Merge the identifiers together

### Cancer Cell Line Encyclopedia
```{r, warning=FALSE, message=FALSE}

effect[, 1] = ensembl
pvalue[, 1] = ensembl

effect = merge(map, effect, by.x='ENSEMBL', by.y='genes')
pvalue = merge(map, pvalue, by.x='ENSEMBL', by.y='genes')

install.packages("writexl")
library("writexl")
filename = "~/Data/RNASeq/CCLE/Mapped_Processed_CCLE.xlsx"
sheets = list("EffectSize"=effect, "PValue"=pvalue)
write_xlsx(sheets, filename)

# Get gene identifiers
geneIDs = df[, c("ENSEMBL", "ENTREZID", "Gene Name")]

# Get raw data
drops = c("ENSEMBL", "ENTREZID", "SYMBOL", "Gene Name")
df = df[, -c(1:4)]

# Get Tissue and Cell Line names
cols = colnames(df)
cols = cols[5:length(cols)]
cols = as.data.frame(cols)

library(tidyr)
cols = separate(data=cols, col=cols, into=c("Tissue", "Cell Line"), sep=", ")
```

### Khao et al., 2020
```{r, warning=FALSE, message=FALSE}
library(tools)

map[, 1] = tolower(map[, 1])
substr(map[, 1], 1, 1) = toupper(substr(map[, 1], 1, 1))

final_df = merge(map, effect, by.x='SYMBOL', by.y='Gene (n=2,020)')
print(final_df)
```

### Lu et al., 2019
```{r, warning=FALSE, message=FALSE}
library(tools)

map[, 1] = tolower(map[, 1])
substr(map[, 1], 1, 1) = toupper(substr(map[, 1], 1, 1))
print(map)
```

```{r, warning=FALSE, message=FALSE}
final_effect = merge(map, effect, by.x='SYMBOL', by.y='Majority Gene name')
final_pvalue = merge(map, pvalue, by.x='SYMBOL', by.y='Majority Gene name')

print(final_effect)
```

# Save all data as Excel file

## Cancer Cell Line Encyclopedia
```{r, warning=FALSE, message=FALSE}
install.packages("xlsx")
library("xlsx")
filename = ""
write.xlsx(geneIDs, filename, 
           sheetName="GeneID",
           col.names=TRUE, row.names=FALSE)
write.xlsx(df, filename, 
           sheetName="Eset", 
           col.names=TRUE, row.names=FALSE)
write.xlsx(cols, filename, 
           sheetName="CellIDs",
           col.names=TRUE, row.names=FALSE)
```

## Khao et al., 2020
```{r, warning=FALSE, message=FALSE}
install.packages('openxlsx') # Install just once
library(openxlsx) # Load on each new session
filename = "~/Data/RNASeq/Dou/MappedDE.xlsx"
write.xlsx(final_df, filename, 
           sheetName="DE",
           col.names=TRUE, row.names=FALSE)
```